### DISCUSSION TITLE
Http Proxy Token Expiration?

### DISCUSSION THREAD

**Discussion by zachhutchins** [mia-platform/community#386] [Q&A]
### What product are you having troubles with?

Console

### What Console version are you using?

v12.0.2

### Description

I am having a unique issue with the http proxy manager v1.6.2 where it is returning a 401 with a new token. For context we are using `api-gateway` which is using `authorization service` and `jwt-token-validator` to validate tokens.

This only happens sometimes, if I hit the endpoints it's trying to hit manually using the token this produces things are working fine. So, this leaves me to believe it's some bug inside the proxy. Without insight into the codebase, it's hard to tell what some of these logs might mean like the following. The service the proxy is calling I can see logs being produce that suggest the token is making it there just fine, but I am curious how we get a log that says token expired when it is brand new.
```
{
  "level": 40,
  "msg": "token expired",
  "reqId": "3511488f-6050-474a-ab86-faa4e8cf2da3",
  "targetBaseURL": "<target>",
  "time": 1707243737029
}
```

```
{
  "host": {
    "hostname": "http-proxy-manager"
  },
  "http": {
    "request": {
      "method": "POST",
      "userAgent": {
        "original": "curl/7.88.1"
      }
    }
  },
  "level": 10,
  "msg": "incoming request",
  "reqId": "3511488f-6050-474a-ab86-faa4e8cf2da3",
  "time": 1707243736686,
  "url": {
    "path": "/somepath/"
  }
}
{
  "level": 30,
  "msg": "token not found in local cache",
  "reqId": "3511488f-6050-474a-ab86-faa4e8cf2da3",
  "time": 1707243736686
}
{
  "level": 30,
  "msg": "requesting new access token from token issuer",
  "reqId": "3511488f-6050-474a-ab86-faa4e8cf2da3",
  "time": 1707243736686,
  "tokenIssuerURL": "<Some issuer>"
}
{
  "level": 20,
  "msg": "successfully retrieved access token from token issuer",
  "reqId": "3511488f-6050-474a-ab86-faa4e8cf2da3",
  "targetUrl": "<SomeTarget>",
  "time": 1707243736971,
  "tokenExpiration": "2024-02-06T19:22:16.971408773Z"
}
{
  "level": 30,
  "msg": "handling request",
  "reqId": "3511488f-6050-474a-ab86-faa4e8cf2da3",
  "requestURI": "/somepath/",
  "time": 1707243736971
}
{
  "destinationUrl": "<some destination>",
  "level": 20,
  "msg": "destination URL of the request",
  "reqId": "3511488f-6050-474a-ab86-faa4e8cf2da3",
  "time": 1707243736971
}
{
  "RequestContentLength": 5,
  "RequestHeaders": {
    "Accept": [
      "*/*"
    ],
    "Authorization": [
      "Bearer <TOKEN>"
    ],
    "Content-Length": [
      "5"
    ],
    "Content-Type": [
      "application/x-www-form-urlencoded"
    ]
  },
  "RequestUrl": {
    "Scheme": "https",
    "Opaque": "",
    "User": null,
    "Host": "<Some Host>",
    "Path": "<somepath>",
    "RawPath": "",
    "OmitHost": false,
    "ForceQuery": false,
    "RawQuery": "",
    "Fragment": "",
    "RawFragment": ""
  },
  "level": 20,
  "msg": "request to forward prepared",
  "reqId": "3511488f-6050-474a-ab86-faa4e8cf2da3",
  "time": 1707243736971
}
{
  "level": 40,
  "msg": "token expired",
  "reqId": "3511488f-6050-474a-ab86-faa4e8cf2da3",
  "targetBaseURL": "<target>",
  "time": 1707243737029
}
{
  "host": {
    "hostname": "http-proxy-manager"
  },
  "http": {
    "request": {
      "method": "POST",
      "userAgent": {
        "original": "curl/7.88.1"
      }
    },
    "response": {
      "statusCode": 401,
      "body": {
        "bytes": 37
      }
    }
  },
  "level": 30,
  "msg": "request completed",
  "reqId": "3511488f-6050-474a-ab86-faa4e8cf2da3",
  "responseTime": 342,
  "time": 1707243737029,
  "url": {
    "path": "/somepath/"
  }
}
```

### Actual Outcome

401 UNAUTHORIZED POST

### Expected Outcome

200

**Comment by lucascanna** [comment]
Hi @zachhutchins,

Thanks for reaching us out for this issue!

The "token expired" log is caused by a conservative behaviour of the HTTP Proxy Manager: if the proxied endpoint responds with a 401 or 403 status code, the HTTP Proxy Manager assumes that it is because the token is expired (without even checking the expiration), therefore in such cases it does two things:
- logs the "token expired" string 
- deletes the token from the in-memory cache

Therefore, in your case, it is very likely that your proxied endpoint is responding with a 401 status code (probably not because the token is expired), so you find the "token expired" log for the reason above. 

While we would like to keep the conservative action of deleting the token from the cache in case the proxied endpoint responds a 401 or 403 status code, we can adjust the log that is actually misleading.

Does it sounds good for you?
Thanks!

**Reply by zachhutchins** [reply]
That would be helpful. 

The behavior that is odd though is when I recreate the request the proxy logs its making, I get a 200. For example, this log below produced a 401 in the proxy.

```
{
  "RequestContentLength": 5,
  "RequestHeaders": {
    "Accept": [
      "*/*"
    ],
    "Authorization": [
      "Bearer <TOKEN>"
    ],
    "Content-Length": [
      "5"
    ],
    "Content-Type": [
      "application/x-www-form-urlencoded"
    ],
    "L5d-Client-Id": [
      "x.y.z.identity.linkerd.cluster.local"
    ],
    "L5d-Dst-Canonical": [
      "http-proxy-manager.ns.svc.cluster.local:80"
    ],
    "User-Agent": [
      "curl/7.88.1"
    ],
    "X-Forwarded-For": [
      ""
    ]
  },
  "RequestUrl": {
    "Scheme": "https",
    "Opaque": "",
    "User": null,
    "Host": "host.com",
    "Path": "/x/y/z/",
    "RawPath": "",
    "OmitHost": false,
    "ForceQuery": false,
    "RawQuery": "",
    "Fragment": "",
    "RawFragment": ""
  },
  "level": 20,
  "msg": "request to forward prepared",
  "reqId": "34f531b8-eb77-4982-81bd-92643f0dc904",
  "time": 1707319937495
}
```

This curl recreation produces a 200, keep in mind I am using the same token the log shows it is using. My only guess here is the log is missing information that is causing it to fail or is misleading.

`curl https://host.com/x/y/z/ -d "hello" -H "authorization: Bearer <TOKEN>" -H "Content-Length: 5" -H "Content-Type: application/x-www-form-urlencoded" -H "L5d-Client-Id: x.y.z.identity.linkerd.cluster.local" -H "L5d-Dst-Canonical: http-proxy-manager.ns.svc.cluster.local:80" -H "X-Forwarded-For: " -H "Accept: */*" -H "User-Agent: curl/7.88.1" -v`

**Reply by lucascanna** [reply]
The elements of the request that are not logged by the proxy are method and request body. Is it possible that your curl recreation is not considering these two elements?

In the logs I see "Content-Length: 5", so I assume that the request to the proxy does contain a request body and, consequently, it is likey that it is not a GET request as you are trying to reproduce in the curl

**Reply by zachhutchins** [reply]
curl by default makes the request a POST when the -d option is used, so its defiantly getting the correct Method and Data. When you put the -v flag on it will report this message telling you so `Note: Unnecessary use of -X or --request, POST is already inferred.`

On the far end the network call trace is the `core/jwt-token-validator:1.1.0`, which is failing with this odd log. I am assuming it means it's using a key from the cached key set. Not sure why a token would be in there.
```
{
  "error": {
    "message": "token found in cache jwks"
  },
  "level": 50,
  "msg": "JWT is not valid.",
  "reqId": "63fbc2c7-e378-483a-9fc5-3db9aa110591",
  "time": 1707322906486
}
```

That log does not show up when I do the curl manually. We are using `envoyproxy/envoy:v1.27-latest`

**Reply by lucascanna** [reply]
You are right. I have read your curl too fast :)

It not clear to me whether the `jwt-token-validator` is the service the proxy manager forwards the requests to (therefore it is after the proxy-manager in your call trace) or it is the service that is used by the `authorization-service` to verify the token (therefore it is before the proxy-manager in your call trace).

In other words, which of the following is your HTTP call trace?
1) api-gateway -> authorization-service -> jwt-token-validator -> proxy-manager -> other-service
2) api-gateway -> authorization-service -> proxy-manager -> jwt-token-validator

**Reply by zachhutchins** [reply]
Neither, proxy-manager -> api-gateway -> authorization-service -> jwt-token-validator

We are using the proxy manager to produce a token so it can be used by the api-gateway.

One interesting other note that we have uncovered is according to envoy, the requests being made to the authorization service appear to be the same request.

```
// Produces 401 -> Call to authorization service
':authority', 'api-gateway.ns'
':path', '/auth/x/y/'
':method', 'POST'
':scheme', 'http'
'content-length', '0'
'authorization', 'Bearer <TOKEN>'
'x-forwarded-host', 'api-gateway.ns
'scheme', 'http'
'x-real-ip', '10.42.34.201'
'x-original-uri', '/x/y/'
'original-request-uri', '/x/y/'
'original-request-method', 'POST'
'isbackoffice', '0'
'x-envoy-internal', 'true'
'x-forwarded-for', '10.42.25.98'
'x-envoy-expected-rq-timeout-ms', '10000'

// Produces 200 -> Call to authorization service
':authority', 'api-gateway.ns'
':path', '/auth/x/y/'
':method', 'POST'
':scheme', 'http'
'content-length', '0'
'authorization', 'Bearer <TOKEN>'
'x-forwarded-host', 'api-gateway.ns'
'scheme', 'http'
'x-real-ip', '10.42.29.163'
'x-original-uri', '/x/y/'
'original-request-uri', '/x/y/'
'original-request-method', 'POST'
'isbackoffice', '0'
'x-envoy-internal', 'true'
'x-forwarded-for', '10.42.25.98'
'x-envoy-expected-rq-timeout-ms', '10000'

```

**Reply by lucascanna** [reply]
Ok, based on your inputs, it could be that the authorization-service is not forwarding the `Authorization` header to the jwt-token-validator.

Can you tell me the values of the environment variables HEADERS_TO_PROXY and AUTHORIZATION_HEADERS_TO_PROXY configured in the authorization-service?

**Reply by alexevansvu** [reply]
For the authorization-service, 
HEADERS_TO_PROXY is set to "authorization, client-type, host, x-forwarded-host"
AUTHORIZATION_HEADERS_TO_PROXY is set to "authorization"

**Reply by zachhutchins** [reply]
@Lucascanna If it is not forwarding the authorization header, I would expect the service to never work. I suspect that it's something unique that is coming from the proxy that is causing it to be the only one that seems to fail, something the logs are just not picking up. Also, this is a "sticky" flickering issue, once it fails it will continue to fail. If we restart services the proxy will work fine, it usually once it's been up for a while do things seem to break.  In that trace we restart everything, it's not even clear which peace is the real culprit.

**Reply by lucascanna** [reply]
@zachhutchins you are right! The hypotesis of the authorization-service is not valid since the call that does not pass through the proxy-manager is working.

Based on the fact that "once it fails it will continue to fail" and the restart of the services make it work again, the problem could be with the caching of tokens in the proxy-manager. Basically, the service caches the authorization token until the expiration date. The cache is in-memory, thus a restart of the service erases the cache.

A possible hypotesys could be that, for some reasons, the verification of the expiration date is not working as expected in the proxy-manager and it continues to use expired tokens. Once the expired token reaches the "jwt-token-validator", it fails producing that odd log above (that I agree is misleading and we need to fix).

In order to validate this hypotesys and understand the reason why the expiration date verification is not working, I would need to know whether your JWT has the "expires_in" field and an example of its value. This is beacause the expiration checks are based on that field.
Moreover, since this checks also use [golang time.Now](https://pkg.go.dev/time@go1.21.0#Now), you should also verify that the system clock of the enviornment where the proxy-manager is running is working fine.

**Reply by zachhutchins** [reply]
So out of curiosity I wrote up a proxy that captures traffic and logs it, headers, cookies, path, method, essentially the entire http request. Both requests match identical. Is there any reason this the jwt-token-validator source is not open? Its lack of logs and my lack of source make it more difficult to debug.

Also, we need to keep in mind that the token being used is the very same as the one being used that passes. If it did not have that field wouldn't both requests fail?

**Reply by zachhutchins** [reply]
@Lucascanna So I took the step to also make a quick token validator myself and it seems to be working where the token validator from the marketplace does not.

I am not sure if there is any more information I can provide, unfortunately the logs from the token validator are not telling me what about the request caused the failure. I can say the requests going into it from a http perspective appear to be the same, which makes this a mystery to me, without the source I can't pop up an instance with a debugger or add more logs to figure out the underlying issue.

**Reply by lucascanna** [reply]
Hi @zachhutchins! 
Thanks for your further insights. It seems to be a tricky issue and it is worth it to debug it together in a call. Our customer success team has already ingaged with your team to organize the meeting. We apologies for the inconvinience this may have caused to you and your team so far.

**Reply by zachhutchins** [reply]
@Lucascanna I agree this one is tricky, appreciate you reaching out and setting up a meeting!
